src/cpp/src/continuous_batching_pipeline.cpp

```diff
@@ -210,6 +218,48 @@ class ContinuousBatchingPipeline::Impl {
             }
         }
 
+        static size_t step_count = 0;
+
+        CacheStateDumper dumper(CacheStateDumper::get_run_id_for_generation_step(step_count, "before_eviction"));
+        dumper.dump_cache_state(*m_scheduler, m_requests);
+
+        // evict unimportant blocks from KV cache, if requested
+        if (m_scheduler->get_config().use_cache_eviction) {
+            auto sequence_attention_scores = m_model_runner->get_last_attention_scores();
+            for (const auto& seq_id_and_attention_scores : sequence_attention_scores) {
+                auto seq_id = seq_id_and_attention_scores.first;
+                const auto& attention_scores_for_all_decoder_layers = seq_id_and_attention_scores.second;
+                std::cout << "VSHAMPOR: starting eviction for seq_id " << seq_id << std::endl;
+                if (m_seq_group_id_to_cache_eviction_algo_map.find(seq_id) == m_seq_group_id_to_cache_eviction_algo_map.end()) {
+                    auto num_decoder_layers = attention_scores_for_all_decoder_layers.size();
+                    auto cache_eviction_config = CacheEvictionConfig();
+                    m_seq_group_id_to_cache_eviction_algo_map[seq_id] = CacheEvictionAlgorithm(cache_eviction_config, num_decoder_layers);
+                }
+                auto& cache_eviction_algo = m_seq_group_id_to_cache_eviction_algo_map[seq_id];
+                auto logical_blocks_to_evict = cache_eviction_algo.get_logical_block_indices_to_evict(attention_scores_for_all_decoder_layers);
+                std::cout << "VSHAMPOR: for decoder layer 0, evicting logical blocks ";
+                for (auto idx : logical_blocks_to_evict[0]) std::cout << idx << " ";
+                std::cout << "\n" << std::endl;
+                m_scheduler->free_blocks_from_sequence(seq_id, logical_blocks_to_evict);
+
+//                auto seq_it = std::find_if(m_requests.begin(), m_requests.end(), [seq_group_id](const SequenceGroup::Ptr& val) { return val->get_request_id() == seq_group_id; });
+//                OPENVINO_ASSERT(seq_it != m_requests.end(), "could not find sequence group ", seq_group_id);
+//                auto sequence_group_ptr = *seq_it;
+
+                auto seq_group_ptr_it = std::find_if(m_requests.begin(), m_requests.end(), [seq_id](const SequenceGroup::Ptr& val) { return val->has_sequence_with_id(seq_id); });
+                OPENVINO_ASSERT(seq_group_ptr_it != m_requests.end(), "could not find sequence group with sequence ", seq_id);
+                auto seq_group_ptr = *(seq_group_ptr_it);
+                auto sequence = seq_group_ptr->get_sequence_by_id(seq_id);
+
+                // Assuming that the evicted blocks are always full (since they by design are only selected from intermediate-age blocks)
+                sequence->register_token_eviction(logical_blocks_to_evict[0].size() * m_scheduler->get_config().block_size);
```

Comment:
do we need to perform re-rotation of "right" KV blocks after evicted ones?

Reply:
Not necessarily, although this is supposed to increase the accuracy. As mentioned before, let's evaluate the current state first (accuracy-wise).